### PR TITLE
applications: nrf5340_audio: Update controller version

### PR DESCRIPTION
**Pull request by erikrobstad**
Update controller to version 3424.

Some accompanying changes to
audio_datapath.c:
- Use counters correctly for drift
and presentation compensation, to avoid
stressing the audio PLL too much.
- Only do small adjustments for drift
compensation, to avoid stressing the
audio PLL too much.
- Adjust sdu_ref_us when doing drift
compensation using BT_LL_ACS_NRF53.

OCT-2807

### CODE REVIEW

File: applications/nrf5340_audio/src/audio/audio_datapath.c [hunk 1]
```diff
@@ -72,18 +73,19 @@ LOG_MODULE_REGISTER(audio_datapath, CONFIG_AUDIO_DATAPATH_LOG_LEVEL);
...
-/* 3000 us to allow BLE transmission and (host -> HCI -> controller) */
-#define JUST_IN_TIME_US		  (CONFIG_AUDIO_FRAME_DURATION_US - 3000)
-#define JUST_IN_TIME_THRESHOLD_US 1500
+/* 4000 us to allow BLE transmission and (host -> HCI -> controller) */
+#define JUST_IN_TIME_US		  (CONFIG_AUDIO_FRAME_DURATION_US - 4000)
+#define JUST_IN_TIME_THRESHOLD_US 2000
 
 /* How often to print underrun warning */
 #define UNDERRUN_LOG_INTERVAL_BLKS 5000
+/* clang-format on */
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/audio_datapath.c:88]: Why do we need to turn off Clang here?

File: applications/nrf5340_audio/src/audio/audio_datapath.c [hunk 2]
```diff
@@ -165,14 +167,19 @@ static size_t test_tone_size;
...
  *
- * @return	err_us	Error in microseconds.
+ * @return	Error in microseconds (err_us).
  */
 static int32_t err_us_calculate(uint32_t sdu_ref_us, uint32_t frame_start_ts)
 {
 	bool err_neg = false;
+
+	if (IS_ENABLED(CONFIG_BT_LL_ACS_NRF53) && IS_ENABLED(CONFIG_TRANSPORT_BIS)) {
+		sdu_ref_us += CONFIG_AUDIO_FRAME_DURATION_US;
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/audio_datapath.c:180]: CONFIG_AUDIO_FRAME_DURATION_US will work just fine, but do we have another define/variable which comes from the BLE/ISO side and not audio? Discuss

File: applications/nrf5340_audio/src/audio/audio_datapath.c [hunk 3]
```diff
@@ -165,14 +167,19 @@ static size_t test_tone_size;
...
+ * @param	sdu_ref_us	Timestamp for SDU.
  * @param	frame_start_ts	Timestamp for I2S.
  *
- * @return	err_us	Error in microseconds.
+ * @return	Error in microseconds (err_us).
  */
 static int32_t err_us_calculate(uint32_t sdu_ref_us, uint32_t frame_start_ts)
 {
 	bool err_neg = false;
+
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/audio_datapath.c:178]: Please add a comment here. Something similar to what you have in the commit message.

File: applications/nrf5340_audio/src/audio/audio_datapath.c [hunk 4]
```diff
@@ -269,8 +278,12 @@ static void audio_datapath_drift_compensation(uint32_t frame_start_ts)
 			return;
 		}
 
+		ctrl_blk.drift_comp.ctr = 0;
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/audio_datapath.c:281]: Discuss these zeroings

File: applications/nrf5340_audio/src/audio/audio_datapath.c [hunk 5]
```diff
@@ -269,8 +278,12 @@ static void audio_datapath_drift_compensation(uint32_t frame_start_ts)
...
+
 		int32_t err_us = err_us_calculate(ctrl_blk.previous_sdu_ref_us, frame_start_ts);
 
+		/* Use asymptotic to get smaller corrections */
+		err_us /= 2;
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/audio_datapath.c:286]: This is effectively the proportional part of a P-regulator. 
This is usually done by multiplication.
```suggestion
		err_us = err_us * REGULATOR_K_P;
```
Then `#define REGULATOR_K_P (0.5)`

**Review by koffes — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/audio/audio_datapath.c [hunk 6]
```diff
@@ -375,6 +389,8 @@ static void audio_datapath_presentation_compensation(uint32_t recv_frame_ts_us,
 			break;
 		}
 
+		ctrl_blk.pres_comp.ctr = 0;
+
 		pres_adj_us = ctrl_blk.pres_comp.sum_err_dly_us / PRES_COMP_NUM_DATA_PTS;
 		if ((pres_adj_us >= (BLK_PERIOD_US / 2)) || (pres_adj_us <= -(BLK_PERIOD_US / 2))) {
```
Comment (gWacey) [inline applications/nrf5340_audio/src/audio/audio_datapath.c:395]: What are the `2`'s here and at line 194?
Could these be a `#define`?
There are a lot of these `2`'s throughout the code, can they all be one or more `#defines` for ease of maintenance, or have a single define of `(BLK_PERIOD_US / 2)`?

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 2 above)
Comment (alexsven) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:180]: No, we looked at it yesterday, and this is the only one that contains 10000 us, it is also heavily used already in this file

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 5 above)
Comment (alexsven) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:286]: Wouldn't you then require a float operation, which takes more time?

File: applications/nrf5340_audio/src/audio/audio_datapath.c [hunk 7]
```diff
@@ -287,18 +300,18 @@ static void audio_datapath_drift_compensation(uint32_t frame_start_ts)
 			return;
 		}
 
+		ctrl_blk.drift_comp.ctr = 0;
+
 		int32_t err_us = err_us_calculate(ctrl_blk.previous_sdu_ref_us, frame_start_ts);
 
-		/* Use asymptotic correction with small errors */
+		/* Use asymptotic to get smaller corrections */
```
Comment (alexsven) [inline applications/nrf5340_audio/src/audio/audio_datapath.c:307]: This sentence doesn't make sense anymore, suggest changing it back to the way it was. 'Asymptotic' is an adjective

File: applications/nrf5340_audio/src/audio/audio_datapath.c [hunk 8]
```diff
@@ -165,14 +167,19 @@ static size_t test_tone_size;
...
  */
 static int32_t err_us_calculate(uint32_t sdu_ref_us, uint32_t frame_start_ts)
 {
 	bool err_neg = false;
+
+	if (IS_ENABLED(CONFIG_BT_LL_ACS_NRF53) && IS_ENABLED(CONFIG_TRANSPORT_BIS)) {
+		sdu_ref_us += CONFIG_AUDIO_FRAME_DURATION_US;
+	}
+
 	int64_t total_err = (sdu_ref_us - frame_start_ts);
```
Comment (alexsven) [inline applications/nrf5340_audio/src/audio/audio_datapath.c:183]: As discussed, might have to change this to an ``int``

**Review by alexsven — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 1 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:88]: I will double check this

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 5 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:286]: #define REGULATOR_DIV_FACTOR 2

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 7 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:307]: Remove these comments

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 8 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:183]: Cast uint32_t to (int64_t)

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 1 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:88]: Removed, as it is not needed

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 3 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:178]: Added comment

File: applications/nrf5340_audio/src/audio/audio_datapath.c [hunk 9]
```diff
@@ -287,18 +300,18 @@ static void audio_datapath_drift_compensation(uint32_t frame_start_ts)
 			return;
 		}
 
+		ctrl_blk.drift_comp.ctr = 0;
+
 		int32_t err_us = err_us_calculate(ctrl_blk.previous_sdu_ref_us, frame_start_ts);
 
-		/* Use asymptotic correction with small errors */
+		/* Use asymptotic to get smaller corrections */
 		err_us /= 2;
```
[the remark this answers is not in the document]
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:308]: Done, made a define

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 6 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:395]: This is a good point, but we need to keep this PR as small as possible, so need to do this change in another PR

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 7 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:307]: Done

File: applications/nrf5340_audio/src/audio/audio_datapath.c (hunk 8 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/audio_datapath.c:183]: Done

**Review by koffes — APPROVED** [review]

**Review by gWacey — APPROVED** [review]

**Review by koffes — APPROVED** [review]

**Review by alexsven — APPROVED** [review]

File: doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst [hunk 10]
```diff
@@ -9,6 +9,21 @@ Changelog
...
+
+Changes
+=======
+
+* Improved robustness of the timestamps (the SDU references) received together with the ISO data.
+* Changed offset of the timestamps (the SDU references) received together with the ISO data to get correct presentation delay.
+* Freed more pins for front-end module (FEM) control.
+* Improved ACL and BIS coexistence on the same device.
+* Fixed an issue where the peripheral did not always accept timing values from the same central.
+* Fixed an issue where the PA sync skip parameter were not used.
```
Comment (greg-fer) [inline doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst:25]:
```suggestion
* Fixed an issue where the PA sync skip parameter was not used.
```

File: doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst [hunk 11]
```diff
@@ -9,6 +9,21 @@ Changelog
...
+* Fixed an issue where it was not possible to create BIG sync after terminating pending BIG sync.
```
Comment (greg-fer) [inline doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst:26]:
```suggestion
* Fixed an issue where it was not possible to create BIG sync after terminating a pending BIG sync.
```

**Review by greg-fer — CHANGES_REQUESTED** [review]

File: doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst (hunk 10 above)
Comment (erikrobstad) [reply doc/nrf/libraries/bin/bt_ll_acs_nrf53/CHANGELOG.rst:25]: Changed

**Review by greg-fer — APPROVED** [review]

**Pull request merged**